File: utils/read_binary_data/matlab/read_single_field_binary.m
% -
%
% SPDX-FileCopyrightText: Pedro Costa and the CaNS contributors
% SPDX-License-Identifier: MIT
%
% -
%
% setting up some parameters
%
precision = 'double';     % precision of the real-valued data
r0 = [0.,0.,0.];    % domain origin
%
% read geometry file
%
geofile  = "geometry.out";
data = dlmread(geofile);
ng   = data(1,:);
%
% read grid
%
f   = fopen('grid_x.bin');
grid_x = fread(f,[ng(1),4],precision);
fclose(f);
xp = r0(1) + grid_x(:,3)'; % centered  x grid
xu = r0(1) + grid_x(:,4)'; % staggered x grid
f   = fopen('grid_y.bin');
grid_y = fread(f,[ng(2),4],precision);
fclose(f);
yp = r0(2) + grid_y(:,3)'; % centered  y grid
yv = r0(2) + grid_y(:,4)'; % staggered y grid
f   = fopen('grid_z.bin');
grid_z = fread(f,[ng(3),4],precision);
fclose(f);
zp = r0(3) + grid_z(:,3)'; % centered  z grid
zw = r0(3) + grid_z(:,4)'; % staggered z grid
%
% read checkpoint binary file
%
filenamei = input("Name of the binary file written by CaNS (e.g. vex_fld_0000000.bin)]: ")
if isempty(filenamei)
end
iskipx      = input("Data saved every (ix, iy, iz) points. Value of ix? [1]: ")
if isempty(iskipx)
    iskipx = 1
end
iskipy      = input("Data saved every (ix, iy, iz) points. Value of iy? [1]: ")
if isempty(iskipy)
    iskipy = 1
end
iskipz      = input("Data saved every (ix, iy, iz) points. Value of iz? [1]: ")
if isempty(iskipz)
    iskipz = 1
end
iskip       = [iskipx,iskipy,iskipz]
n           = floor((ng-1)./iskip)+1
f = fopen(filenamei);
fld = fread(f,prod(n),precision);
fclose(f);
if numel(fld) ~= prod(n)
    error("expected %d values for the requested skip, found %d",prod(n),numel(fld))
end
data = reshape(fld,[n(1),n(2),n(3)]);

File: utils/read_binary_data/python/read_single_field_binary.py
# -
#
# SPDX-FileCopyrightText: Pedro Costa and the CaNS contributors
# SPDX-License-Identifier: MIT
#
# -
#!/usr/bin/env python
def read_single_field_binary(data_dir,filenamei,iskip):
    import os
    import numpy as np
    #
    # setting up some parameters
    #
    iprecision = 8            # precision of the real-valued data
    r0 = np.array([0.,0.,0.]) # domain origin
    precision  = 'float64'
    if(iprecision == 4): precision = 'float32'
    #
    # read geometry file
    #
    geofile  = data_dir+"/geometry.out"
    geo = np.loadtxt(geofile, comments = "!", max_rows = 2)
    ng = geo[0,:].astype('int')
    #
    # read grid
    #
    grids = []
    for axis, n, offset in zip(["x","y","z"],ng,r0):
        grid = np.fromfile(os.path.join(data_dir,"grid_"+axis+".bin"),dtype=precision)
        grid = np.reshape(grid,(n,4),order='F')
        grids.append((offset + grid[:,2], offset + grid[:,3]))
    #
    # split centered and staggered grids
    #
    xp = grids[0][0] # centered grid
    xu = grids[0][1] # staggered grid
    yp = grids[1][0] # centered grid
    yv = grids[1][1] # staggered grid
    zp = grids[2][0] # centered grid
    zw = grids[2][1] # staggered grid
    #
    # read binary file
    #
    iskip       = np.asarray(iskip,dtype=int)
    n           = ((ng[:]-1)//iskip[:] + 1).astype(int)
    fld         = np.fromfile(data_dir+"/"+filenamei,dtype=precision)
    if(fld.size != np.prod(n)):
        raise ValueError("expected {} values for ng={} and iskip={}, found {}".format(np.prod(n),ng,iskip,fld.size))
    data        = np.reshape(fld,(n[0],n[1],n[2]),order='F')
    #
    # reshape grid
    #
    xp = xp[0:ng[0]:iskip[0]]
    yp = yp[0:ng[1]:iskip[1]]
    zp = zp[0:ng[2]:iskip[2]]
    xu = xu[0:ng[0]:iskip[0]]
    yv = yv[0:ng[1]:iskip[1]]
    zw = zw[0:ng[2]:iskip[2]]
    return data, xp, yp, zp, xu, yv, zw

if __name__ == "__main__":
    import numpy as np
    filenamei   = input("Name of the binary file written by CaNS (e.g. vex_fld_0000000.bin)]: ")
    iskipx      = input("Data saved every (ix, iy, iz) points. Value of ix? [1]: ") or "1"
    iskipy      = input("Data saved every (ix, iy, iz) points. Value of iy? [1]: ") or "1"
    iskipz      = input("Data saved every (ix, iy, iz) points. Value of iz? [1]: ") or "1"
    iskip       = np.array([iskipx,iskipy,iskipz]).astype(int)
    data, xp, yp, zp, xu, yv, zw = read_single_field_binary("./",filenamei,iskip)
